src/core/lib/security/credentials/tls/tls_utils.h

```diff
@@ -46,6 +50,20 @@ absl::string_view GetAuthPropertyValue(grpc_auth_context* context,
 std::vector<absl::string_view> GetAuthPropertyArray(grpc_auth_context* context,
                                                     const char* property_name);
 
+struct EVP_PKEYDeleter {
+  void operator()(EVP_PKEY* pkey) const { EVP_PKEY_free(pkey); }
+};
+struct BIO_Deleter {
+  void operator()(BIO* bio) const { BIO_free(bio); }
+};
+struct X509_Deleter {
+  void operator()(X509* x509) const { X509_free(x509); }
+};
```

Comment:
As we add more features interacting with OpenSSL/BoringSSL, I think there is a need for creating some helper functions/data structures for OpenSSL/BoringSSL. This could be a good starting point.

On the other side, I am still not sure if a change like this would be an overkill for such kind of task. In gRPC, creating an object and deleting it later is a common pattern, e.g. `gpr_malloc` and `gpr_free`. Looking at the code again, it seems remembering to `free` the object shouldn't be too bad, and changes like this will cause unnecessary indirections while reading the code. We can always keep it simple, and define such helper functions when deleting such objects really becomes a problem.
@markdroth may I know your opinion about this change?

Reply:
I imagine the commonness of a manual freeing comes from gRPC originally being in C. RAII  is probably the single biggest benefit C++ goes over C. While it's true that, in many functions, it's not too bad, it's a huge benefit across the codebase to be able to do it automatically. Code becomes dramatically less error-prone when you don't need to manually free everything on ever return path.

Reply:
OK, that makes sense. Being able to not free on every return path does provide benefits. 
Yes, originally gRPC has to be in C but now we can use C++ everywhere except some interfaces. 
Let's proceed with the change then. Thanks for the suggestion!